packages/tuono-router/src/components/RouterContext.tsx

```diff
@@ -20,3 +36,88 @@ export function getRouterContext(): React.Context<Router> {
 
   return routerContext
 }
+
+interface RouterContextProviderProps {
+  router: Router
+  children: ReactNode
+  serverSideProps?: ServerRouterInfo
+}
+
+function getInitialLocation(
+  serverSideProps?: ServerRouterInfo,
+): ParsedLocation {
+  if (typeof document === 'undefined') {
+    return {
+      pathname: serverSideProps?.pathname || '',
+      hash: '',
+      href: serverSideProps?.href || '',
+      searchStr: serverSideProps?.searchStr || '',
+      // TODO: Polyfill URLSearchParams
+      search: {},
+    }
+  }
+
+  const { location } = window
+  return {
+    pathname: location.pathname,
+    hash: location.hash,
+    href: location.href,
+    searchStr: location.search,
+    search: Object.fromEntries(new URLSearchParams(location.search)),
+  }
+}
+
+export function RouterContextProvider({
+  router,
+  children,
+  serverSideProps,
+}: RouterContextProviderProps): ReactNode {
+  // Allow the router to update options on the router instance
+  router.update({ ...router.options } as Parameters<typeof router.update>[0])
+
+  const [location, updateLocation] = useState<ParsedLocation>(
+    getInitialLocation(serverSideProps),
+  )
+
+  /**
+   * Listen browser navigation events
+   */
+  useEffect(() => {
+    const updateLocationOnPopStateChange = ({
+      target,
+    }: PopStateEvent): void => {
+      const { location: targetLocation } = target as typeof window
+      const { pathname, hash, href, search } = targetLocation
+
+      updateLocation({
+        pathname,
+        hash,
+        href,
+        searchStr: search,
+        search: Object.fromEntries(new URLSearchParams(search)),
+      })
+    }
+
+    window.addEventListener('popstate', updateLocationOnPopStateChange)
+
+    return (): void => {
+      window.removeEventListener('popstate', updateLocationOnPopStateChange)
+    }
+  }, [updateLocation])
+
+  const RouterContext = getRouterContext()
+
+  const context: RouterContextType = {
+    router,
+    location,
+    updateLocation,
```

Comment:
Since this context wraps the whole application I would consider to wrap the context value in a `React.useMemo`

Reply:
What props might cause re-rendering?

Reply:
probably `children`, also the component is not wrapped in a `React.memo` so any change in the components up in the tree might cause its re-rendering.
I understand this is used in the app entry point and has a limited impact, but better be safe than sorry 😅

Reply:
Definitely!